packages/explorerkit-server/src/server.ts

```diff
@@ -203,67 +88,91 @@ app.post("/decode/accounts", responseDurationMiddleware, async (req: Request, re
-        return res.status(400).json({ error: "Invalid request body" });
-      }
-      let topLevelInstruction = instruction.topLevelInstruction;
-      if (
-        topLevelInstruction.programId === undefined ||
-        topLevelInstruction.encodedData === undefined ||
-        topLevelInstruction.accountKeys === undefined ||
-        !Array.isArray(topLevelInstruction.accountKeys) ||
-        !Array.isArray(instruction.flattenedInnerInstructions)
-      ) {
-        return res.status(400).json({ error: "Invalid request body" });
-      }
-      let flattenedInnerInstructions = instruction.flattenedInnerInstructions;
-      for (let innerInstruction of flattenedInnerInstructions) {
-        if (
-          innerInstruction.programId === undefined ||
-          innerInstruction.encodedData === undefined ||
-          innerInstruction.accountKeys === undefined ||
-          !Array.isArray(innerInstruction.accountKeys)
-        ) {
-          return res.status(400).json({ error: "Invalid request body" });
-        }
-      }
-    }
+  const { data, error } = decodeInstructionsSchema.safeParse(req.body);
+
+  if (error) {
+    return res.status(400).json({ error: "Invalid request body", errors: error.errors });
   }
 
   try {
-    const { instructionsPerTransaction } = req.body as DecodeTransactionsRequestBody;
+    const { instructionsPerTransaction } = data satisfies DecodeTransactionsRequestBody;
+    const allProgramIds = getProgramIds(instructionsPerTransaction);
+    const idls = await loadAllIdls(allProgramIds);
 
-    let allProgramIds = getProgramIds(instructionsPerTransaction);
-    await loadAllIdls(allProgramIds);
+    const decodedTransactions: (TopLevelInstruction[] | null)[] = [];
 
-    let decodedTransactions: (TopLevelInstruction[] | null)[] = [];
-    for (var transactionInstructions of instructionsPerTransaction) {
-      if (transactionInstructions === null) {
+    for (const transactionInstructions of instructionsPerTransaction) {
+      if (!transactionInstructions) {
         decodedTransactions.push(null);
         continue;
       }
-      let decodedTransaction: TopLevelInstruction[] = [];
-      for (var instruction of transactionInstructions) {
+
+      const decodedTransaction: TopLevelInstruction[] = [];
+
+      for (const instruction of transactionInstructions) {
         // First decode top level ix, then all nested ixs
-        let decodedTopLevelInstruction = await decodeInstruction(instruction.topLevelInstruction);
-        let decodedInnerInstruction = [];
-        for (var inner_instruction of instruction.flattenedInnerInstructions) {
-          decodedInnerInstruction.push(await decodeInstruction(inner_instruction));
+        const decodedTopLevelInstruction = await decodeInstruction(idls, instruction.topLevelInstruction);
+        const decodedInnerInstruction = [];
+
+        for (const inner_instruction of instruction.flattenedInnerInstructions) {
```

Comment:
NIT: I know its existing code but maybe we can clean it up
```suggestion
        for (const innerInstruction of instruction.flattenedInnerInstructions) {
```

Reply:
Done